### PR TITLE
Insights: integrate backend insights with compatible extensions insights

### LINKED ISSUE #18397: insights: discover from user/org settings
Today, insights are discovered from "global user settings" only. This happens [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/insights/discovery/discovery.go). You must be an admin to edit the global user settings. It is a cascade like so:

Global user settings -> Organization settings -> User settings (overrides all else).

Insights violates this cascade and the default user assumption by only picking up ones defined in the global user settings.

### Security

Actually making insights get picked up from user/org settings is pretty straightforward, we merely need to update the `Discover` function here to read from all settings on the entire Sourcegraph instance: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/insights/discovery/discovery.go

However, since insights run search queries in the background with no authentication, they have access to search results the user may not have access to. For example, a plain user could add a search insight which can see repositories only visible to a specific team within a company. If we were to naively enable this, I could write insights which extract _some_ information about the repos I don't have access to:

1. whether or not a specific term exists in someone's repos, and how many instances of it.
2. That repos exist under a given org/name. This is most risky on Sourcegraph.com, e.g. someone could check `repo:microsoft-private/foobar` and see that Microsoft has private repositories on Sourcegraph.com (I can neither confirm nor deny if this is true.)

There are some cross-cutting things to consider as well when thinking about:

* Giving repository-level breakdowns of insights
* Add GraphQL API to push data into TimescaleDB/insights store directly ("webhooks") #17220 - which may be describing global (not associated with a repository) insights.

Probably we need to:

1. Discover insights from all org/user settings on the instance (easy.)
2. When querying insights, if per-repo is being queried then confirm the user has access to those repos one-by-one (we should have an existing API to do this given a repo ID.)
3. When querying insights, if NOT per-repo (i.e. a global insight) then confirm the user has that exact insight defined in their user settings. We could do this by comparing the `SeriesID` which is a hash of the insight's important bits.

### PR DESCRIPTION
**Pull request by coury-clark**
Closes #18397 

This will cause the backend to additionally load extensions insights in a new setting object. This is useful to support first integration, and then migration to the backend.

### CODE REVIEW

File: enterprise/internal/insights/discovery/discovery.go [hunk 1]
```diff
@@ -90,11 +99,14 @@ func filterByIds(ids []string, insight []insights.SearchInsight) []insights.Sear
 	filtered := make([]insights.SearchInsight, 0)
 	keys := make(map[string]bool)
 	for _, id := range ids {
+		log15.Info("filter", "key", id)
```
Comment (slimsag) [inline enterprise/internal/insights/discovery/discovery.go:102]:
```suggestion
```

File: enterprise/internal/insights/discovery/discovery.go [hunk 2]
```diff
@@ -90,11 +99,14 @@ func filterByIds(ids []string, insight []insights.SearchInsight) []insights.Sear
...
 		keys[id] = true
 	}
 
 	for _, searchInsight := range insight {
+		log15.Info("checking_filter", "insight", searchInsight.Title, "uuid", searchInsight.ID)
```
Comment (slimsag) [inline enterprise/internal/insights/discovery/discovery.go:107]:
```suggestion
```

File: enterprise/internal/insights/discovery/discovery.go [hunk 3]
```diff
@@ -90,11 +99,14 @@ func filterByIds(ids []string, insight []insights.SearchInsight) []insights.Sear
...
 		if _, ok := keys[searchInsight.ID]; ok {
+			log15.Info("adding", "insight", searchInsight.Title)
```
Comment (slimsag) [inline enterprise/internal/insights/discovery/discovery.go:109]:
```suggestion
```

File: enterprise/internal/insights/discovery/discovery.go (hunk 2 above)
Comment (slimsag) [reply enterprise/internal/insights/discovery/discovery.go:107]: (if these are actually useful, feel free to leave in)

Comment (coury-clark) [reply enterprise/internal/insights/discovery/discovery.go:107]: `git merge` and I had a bad day today. I'll clean this all up tomorrow

File: internal/insights/insights.go [hunk 4]
```diff
@@ -5,12 +5,32 @@ import (
...
 	"github.com/sourcegraph/sourcegraph/internal/database"
 	"github.com/sourcegraph/sourcegraph/internal/database/dbutil"
 	"github.com/sourcegraph/sourcegraph/internal/jsonc"
 )
 
+type Loader interface {
+	LoadAll(ctx context.Context) ([]SearchInsight, error)
+}
+
+type DBLoader struct {
```
Comment (slimsag) [inline internal/insights/insights.go:22]: Need comments on this and Loader, to describe what these concepts are (I can't easily guess what a loader does)

File: internal/insights/insights.go [hunk 5]
```diff
@@ -115,6 +137,84 @@ func GetLangStatsInsights(ctx context.Context, db dbutil.DB, filter SettingFilte
...
+			if err != nil {
+				// this isn't actually a total failure case, we could have partially parsed this dictionary.
+				multi = multierror.Append(multi, err)
+			}
+			results = append(results, temp.Insights()...)
+		}
+	}
+
+	if multi != nil {
+		log15.Error("deserialization errors parsing integrated insights", "error", multi)
```
Comment (slimsag) [inline internal/insights/insights.go:176]:
```suggestion
		log15.Error("insights: deserialization errors parsing integrated insights", "error", multi)
```

Comment (slimsag) [reply internal/insights/insights.go:176]: maybe you want to return the error here as well?

**Review by slimsag — APPROVED** [review]
Looks pretty great to me!

File: internal/insights/insights.go (hunk 5 above)
Comment (coury-clark) [reply internal/insights/insights.go]: In this case, we are silencing the error completely. It's kind of a weird thing, but there is no guarantee these settings are well-formed. So, we will throw out any that aren't well-formed. This is a temporary solution anyway, until we migrate all of these to a database.

**Pull request merged**